text/0000-unicode-idents.md

```diff
@@ -0,0 +1,145 @@
+
+# Motivation
+[motivation]: #motivation
+
+Rust is written by many people who are not fluent in the English language. Using identifiers in ones native language eases writing and reading code for these developers.
+
+The rationale from [PEP 3131] nicely explains it:
+
+> ~~Python~~ *Rust* code is written by many people in the world who are not familiar with the English language, or even well-acquainted with the Latin writing system. Such developers often desire to define classes and functions with names in their native languages, rather than having to come up with an (often incorrect) English translation of the concept they want to name. By using identifiers in their native language, code clarity and maintainability of the code among speakers of that language improves.
+> 
+> For some languages, common transliteration systems exist (in particular, for the Latin-based writing systems). For other languages, users have larger difficulties to use Latin to write their native words.
+
+Additionally some math oriented projects may want to use identifiers closely resembling mathematical writing.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Identifiers include variable names, function and trait names and module names. They start with a letter or an underscore and may be followed by more letters, digits and some connecting punctuation.
+
+Examples of valid identifiers are:
+
+* English language words: `color`, `image_width`, `line2`, `Photo`, `_unused`, ...
+* ASCII words in foreign languages: `die_eisenbahn`, `el_tren`, `artikel_1_grundgesetz`
+* words containing accented characters: `garçon`, `hühnervögel`
+* identifiers in other scripts: `Москва`, `東京`, ...
+
+Examples of invalid identifiers are:
+
+* Keywords: `impl`, `fn`, `_` (underscore), ...
+* Identifiers starting with numbers or "non letters": `42_the_answer`, `third√of7`, `◆◆◆`, ...
+* Emojis: 🙂, 🦀, 💩, ...
+
+Similar Unicode identifiers are normalized: `a1` and `a₁` refer to the same variable. This also applies to accented characters which can be represented in different ways.
+
+To disallow any Unicode identifiers in a project (for example to ease collaboration or for security reasons) limiting the accepted identifiers to ASCII add this lint to the `lib.rs` or `main.rs` file of your project:
+
+```rust
+#![forbid(unicode_idents)]
+```
+
+Some Unicode character look confusingly similiar to each other or even identical like the Latin **A** and the Cyrillic **А**. The compiler may warn you about easy to confuse names in the same scope. If needed (but not recommended) this warning can be silenced with a `#[allow(confusable_unicode_idents)]` annotation on the enclosing function or module.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Identifiers in Rust are based on the [Unicode® Standard Annex #31 Unicode Identifier and Pattern Syntax][TR31]. Rust compilers shall use at least Revision 27 of the standard.
```

Comment:
The revision should be specified exactly. Otherwise, the same identifier in the same Rust version could be valid in one compiler but another.

Reply:
I don't want to lock Rust compilers to a specific Unicode revision. An alternative would be to state the supported Unicode revision for each Rust edition.

Reply:
New versions of Rust should use new versions of Unicode, of course, but any given version of Rust should have an unambiguous definition of identifiers. Your alternative is good.

Here is why the current text is wrong. Let’s say, for example, that revision 28 adds ⟨⍙⟩ to XID_Start, just in time for Rust 1.40.0. Would ⟨⍙abc⟩ be a valid identifier in Rust 1.40.0? It would be valid in compilers that use revision 28, but not in compilers that still use revision 27. Both compilers would be correct and yet disagree about something as basic as identifier validity.

Reply:
FWIW we’ve already been updating Unicode for the standard library (e.g. `char::to_lowercase`), and have https://doc.rust-lang.org/std/char/constant.UNICODE_VERSION.html to indicate which version is in use.

Reply:
UAX 31 should be referenced for information/context, but I think that its *revision* number can be removed entirely. This RFC defines identifier syntax based on the `XID_Start` and `XID_Continue` properties, which are specified exactly for a given *version* of Unicode (e.g. 10.0.0).

To maintain Rust’s stability promise, we need to ensure that:

1. If a given string is a valid identifier in a given version of Rust, it needs to stay a valid identifier in later versions of Rust.
2. If two strings are valid identifiers in a given version of Rust, whether they compare equal after normalization needs to be unchanged in later versions of Rust.

My reading of UAX 31 and 15 is that we can do that and still update the to newer versions of Unicode in newer versions of Rust backward-compatibly without involving Rust editions. We should have some way communicate to users the version mapping, and I think that the existing `std::char::UNICODE_VERSION` (with docs archive like https://doc.rust-lang.org/1.15.0/std/char/constant.UNICODE_VERSION.html) is already satisfactory for that.

This translates 1. above into the requirement [UAX31-R1b Stable Identifier](https://www.unicode.org/reports/tr31/#R1b):

> To meet this requirement, an implementation shall guarantee that identifiers are stable across versions of the Unicode Standard: that is, once a string qualifies as an identifier, it does so in all future versions.

Per https://www.unicode.org/reports/tr31/#Backward_Compatibility I believe that `XID_Start XID_Continue* | "_" XID_Continue+` meets that requirement. (Since XID_Start / XID_Continue include Other_ID_Start / Other_ID_Continue respectively. I’d appreciate if someone could double-check and confirm my understanding.)

Additionally, code points that are not assigned in a given Unicode version cannot be in `XID_Start` or `XID_Continue` in that Unicode version. And https://www.unicode.org/reports/tr15/#Versioning defines:

> It is crucial that Normalization Forms remain stable over time. That is, if a string that does not have any unassigned characters is normalized under one version of Unicode, it must remain normalized under all future versions of Unicode. 

Updating to a newer Unicode does not change the result of `NFKC(X)` or `NFKC(Y)`, so `NFKC(X) = NFKC(Y)` is also unchanged and requirement 2. above is also met.

-----

@dscorbett 

>  Let’s say, for example, that revision 28 adds ⟨⍙⟩ to XID_Start, just in time for Rust 1.40.0. Would ⟨⍙abc⟩ be a valid identifier in Rust 1.40.0?

It would a Unicode version rather than in a revision of UAX 31 that adds it, but yes, if Rust 1.40.0 is the one that upgrades to that version of Unicode (regardless of when that version of Unicode was released), then `⍙abc` would be a valid ident in Rust 1.40.0 but not 1.39.0. This is similar to using new a language feature newly in 1.40.0.

Reply:
Right. It was a rhetorical question to show why specifying the Unicode version exactly is necessary.